README.md
JBlake2 [![CI][1]][2] [![Coverage][3]][4] [![Reproducible Builds][5]][6] [![Version][7]][8] [![Javadocs][9]][10] [![License][11]][12]
=====================================================================================================================================

A pure Java (8+) implementation of BLAKE2 ([RFC 7693][13]).


Getting started
---------------

Maven
```xml
  <dependency>
    <groupId>org.kocakosm</groupId>
    <artifactId>jblake2</artifactId>
    <version>0.5</version>
  </dependency>
```

Gradle
```groovy
  compile 'org.kocakosm:jblake2:0.5'
```

*Note: module name for the Java Module System is `org.kocakosm.jblake2`.*


Usage
-----

Browse [API docs][10] for the most recent release.


Reproducibility
---------------

JBlake2 builds are reproducible since version 0.4. Released artifacts are
verified by the [Reproducible Central][14] project, the Java part of the
[Reproducible Builds][15] effort.


License
-------

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU Lesser General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option) any
later version.

This program is distributed in the hope that it will be useful, but _without any
warranty;_ without even the implied warranty of _merchantability_ or _fitness
for a particular purpose_.

See the [GNU Lesser General Public License][12] for more details.


Contact
-------

kocakosm[@]gmail[dot]com


 [1]: https://github.com/kocakosm/jblake2/actions/workflows/ci.yml/badge.svg
 [2]: https://github.com/kocakosm/jblake2/actions/workflows/ci.yml
 [3]: https://img.shields.io/coveralls/kocakosm/jblake2.svg
 [4]: https://coveralls.io/github/kocakosm/jblake2
 [5]: https://img.shields.io/badge/reproducible_builds-ok-success?labelColor=1e5b96
 [6]: https://github.com/jvm-repo-rebuild/reproducible-central#org.kocakosm:jblake2
 [7]: https://img.shields.io/maven-central/v/org.kocakosm/jblake2.svg
 [8]: https://central.sonatype.com/search?q=g%3Aorg.kocakosm+a%3Ajblake2
 [9]: https://javadoc.io/badge/org.kocakosm/jblake2.svg
 [10]: https://javadoc.io/doc/org.kocakosm/jblake2
 [11]: https://img.shields.io/badge/license-LGPL_v3-4383c3.svg
 [12]: https://www.gnu.org/licenses/lgpl-3.0.txt
 [13]: https://tools.ietf.org/html/rfc7693
 [14]: https://github.com/jvm-repo-rebuild/reproducible-central
 [15]: https://reproducible-builds.org


CHANGELOG
## 0.5 (2023-09-08)
- Blake2#burn() now returns 'this'
- Code cleanup

## 0.4 (2018-12-21)
- Added a script to easily reproduce/verify a release build
- Javadoc improvements
- Code cleanup

## 0.3 (2018-08-10)
- Implemented Blake2s
- Migrated to Java 8
- Fixed module name for the Java Module System
- Made the build process entirely deterministic
- Migrated unit tests to Junit 5
- Javadoc improvements
- Code cleanup

## 0.2 (2017-11-30)
- Javadoc improvements
- Code cleanup

## 0.1 (2017-11-29)
- Initial release


test/org/kocakosm/jblake2/PRNG.java
/*----------------------------------------------------------------------------*
 * This file is part of JBlake2.                                              *
 * Copyright (C) Osman Koçak <kocakosm@gmail.com>                             *
 *                                                                            *
 * This program is free software: you can redistribute it and/or modify it    *
 * under the terms of the GNU Lesser General Public License as published by   *
 * the Free Software Foundation, either version 3 of the License, or (at your *
 * option) any later version.                                                 *
 * This program is distributed in the hope that it will be useful, but        *
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY *
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public     *
 * License for more details.                                                  *
 * You should have received a copy of the GNU Lesser General Public License   *
 * along with this program. If not, see <http://www.gnu.org/licenses/>.       *
 *----------------------------------------------------------------------------*/

package org.kocakosm.jblake2;

import java.util.Random;

/**
 * Utility class to generate random data for tests.
 *
 * @author Osman Koçak
 */
final class PRNG
{
	private static final Random RANDOM = new Random();

	static byte[] nextBytes(int count)
	{
		if (count < 0) {
			throw new IllegalArgumentException();
		}
		byte[] bytes = new byte[count];
		RANDOM.nextBytes(bytes);
		return bytes;
	}

	static int nextInt()
	{
		return RANDOM.nextInt();
	}

	static int nextInt(int startInclusive, int endExclusive)
	{
		if (startInclusive < 0 || endExclusive <= startInclusive) {
			throw new IllegalArgumentException();
		}
		return startInclusive + RANDOM.nextInt(endExclusive - startInclusive);
	}

	static long nextLong()
	{
		return RANDOM.nextLong();
	}

	private PRNG()
	{
		// See "Effective Java" (Item 4)
		throw new AssertionError("Not meant to be instantiated");
	}
}

test/org/kocakosm/jblake2/Base16.java
/*----------------------------------------------------------------------------*
 * This file is part of JBlake2.                                              *
 * Copyright (C) 2017 Osman Koçak <kocakosm@gmail.com>                        *
 *                                                                            *
 * This program is free software: you can redistribute it and/or modify it    *
 * under the terms of the GNU Lesser General Public License as published by   *
 * the Free Software Foundation, either version 3 of the License, or (at your *
 * option) any later version.                                                 *
 * This program is distributed in the hope that it will be useful, but        *
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY *
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public     *
 * License for more details.                                                  *
 * You should have received a copy of the GNU Lesser General Public License   *
 * along with this program. If not, see <http://www.gnu.org/licenses/>.       *
 *----------------------------------------------------------------------------*/

package org.kocakosm.jblake2;

/**
 * Base16 (RFC 4648) decoding.
 *
 * @author Osman Koçak
 */
final class Base16
{
	public static byte[] decode(String hex)
	{
		int len = hex.length();
		if (len % 2 != 0) {
			throw new IllegalArgumentException();
		}
		byte[] data = new byte[len / 2];
		for (int i = 0; i < len; i += 2) {
			int val = Integer.parseInt(hex.substring(i, i + 2), 16);
			data[i / 2] = (byte) (val & 0xFF);
		}
		return data;
	}

	private Base16()
	{
		/* ... */
	}
}


test/org/kocakosm/jblake2/TestVector.java
/*----------------------------------------------------------------------------*
 * This file is part of JBlake2.                                              *
 * Copyright (C) Osman Koçak <kocakosm@gmail.com>                             *
 *                                                                            *
 * This program is free software: you can redistribute it and/or modify it    *
 * under the terms of the GNU Lesser General Public License as published by   *
 * the Free Software Foundation, either version 3 of the License, or (at your *
 * option) any later version.                                                 *
 * This program is distributed in the hope that it will be useful, but        *
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY *
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public     *
 * License for more details.                                                  *
 * You should have received a copy of the GNU Lesser General Public License   *
 * along with this program. If not, see <http://www.gnu.org/licenses/>.       *
 *----------------------------------------------------------------------------*/

package org.kocakosm.jblake2;

import static org.kocakosm.pitaya.util.XArrays.copyOf;

import com.google.gson.annotations.SerializedName;

/**
 * Represents a BLAKE2 test vector.
 *
 * @author Osman Koçak
 */
final class TestVector
{
	@SerializedName("key")
	private final byte[] key;

	@SerializedName("in")
	private final byte[] input;

	@SerializedName("out")
	private final byte[] output;

	TestVector(byte[] key, byte[] input, byte[] output)
	{
		this.key = copyOf(key);
		this.input = copyOf(input);
		this.output = copyOf(output);
	}

	byte[] getKey()
	{
		return copyOf(key);
	}

	byte[] getInput()
	{
		return copyOf(input);
	}

	byte[] getOutput()
	{
		return copyOf(output);
	}
}

verify-build.sh
#!/usr/bin/env bash

#----------------------------------------------------------------------------#
# This file is part of JBlake2.                                              #
# Copyright (C) Osman Koçak <kocakosm@gmail.com>                             #
#                                                                            #
# This program is free software: you can redistribute it and/or modify it    #
# under the terms of the GNU Lesser General Public License as published by   #
# the Free Software Foundation, either version 3 of the License, or (at your #
# option) any later version.                                                 #
# This program is distributed in the hope that it will be useful, but        #
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY #
# or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public     #
# License for more details.                                                  #
# You should have received a copy of the GNU Lesser General Public License   #
# along with this program. If not, see <http://www.gnu.org/licenses/>.       #
#----------------------------------------------------------------------------#

set -eEuo pipefail
IFS=$'\n\t'

readonly ARCH='x86_64'
readonly MIN_BASH_VERSION='4.1'
readonly MIN_DOCKER_VERSION='17.05'
readonly BASE_IMAGE_NAME='amd64/debian:bullseye'
readonly BUILD_IMAGE_NAME='jblake2-build'
readonly ENTRYPOINT_SCRIPT_NAME='entrypoint.sh'

check_bash_version()
{
	if [[ "${BASH_VERSION}" < "${MIN_BASH_VERSION}" ]]; then
		printf 'Error: you need GNU Bash version %s+ to run this script\n' "${MIN_BASH_VERSION}" >&2
		exit 1
	fi
}

check_cpu_architecture()
{
	if ! [[ "$(uname -m)" = "${ARCH}" ]]; then
		printf 'Error: you need an %s CPU to run this script\n' "${ARCH}" >&2
		exit 1
	fi
}

check_docker_version()
{
	if ! [[ -x $(command -v docker) ]]; then
		printf 'Error: you need Docker to run this script\n' >&2
		exit 1
	fi
	if [[ "$(docker version --format '{{.Server.Version}}')" < "${MIN_DOCKER_VERSION}" ]]; then
		printf 'Error: you need Docker version %s+ to run this script\n' "${MIN_DOCKER_VERSION}" >&2
		exit 1
	fi
}

write_entrypoint()
{
	cat <<- 'EOF' > "${ENTRYPOINT_SCRIPT_NAME}"
	#!/usr/bin/env bash
	set -eEuo pipefail
	IFS=$'\n\t'

	readonly MIN_JBLAKE2_VERSION='0.4'
	readonly MAVEN_CENTRAL_URL='https://repo1.maven.org/maven2'

	read_version()
	{
		# disable shellcheck warning on '${project.version}' since it is not a shell variable
		# shellcheck disable=SC2016
		mvn --quiet -Dexec.executable='printf' -Dexec.args='${project.version}' exec:exec
	}

	check_version()
	{
		local -r jblake2_version=${1}
		if [[ "${jblake2_version}" = *-SNAPSHOT ]]; then
			printf 'Error: SNAPSHOT version\n' >&2
			exit 1
		fi
		if [[ "${jblake2_version}" < "${MIN_JBLAKE2_VERSION}" ]]; then
			printf 'Error: JBlake2 is reproducible since version %s\n' "${MIN_JBLAKE2_VERSION}" >&2
			exit 1
		fi
	}

	sha256sum()
	{
		local -r file=${1}
		command sha256sum "${file}" | cut -d ' ' -f 1
	}

	download_jar_from_central()
	{
		local -r jblake2_version=${1}
		curl --silent --location --remote-name \
			"${MAVEN_CENTRAL_URL}/org/kocakosm/jblake2/${jblake2_version}/jblake2-${jblake2_version}.jar"
		printf 'jblake2-%s.jar' "${jblake2_version}"
	}

	build_jar_from_source()
	{
		mvn --quiet clean package -Dmaven.test.skip=true
		ls target/jblake2-*.jar
	}

	main()
	{
		printf 'Reading version...\t\t'
		local -r version=$(read_version)
		printf '%s\n' "${version}"
		check_version "${version}"
		printf 'Downloading jar from central...\t'
		local -r downloaded_jar=$(download_jar_from_central "${version}")
		local -r downloaded_jar_hash=$(sha256sum "${downloaded_jar}")
		printf 'sha256:%s\n' "${downloaded_jar_hash}"
		printf 'Building jar from source...\t'
		local -r built_jar=$(build_jar_from_source)
		local -r built_jar_hash=$(sha256sum "${built_jar}")
		printf 'sha256:%s\n' "${built_jar_hash}"
		if ! [[ "${built_jar_hash}" = "${downloaded_jar_hash}" ]]; then
			printf 'Hashes mismatch: JBlake2 %s could not be verified\n' "${version}"
			exit 1
		fi
		printf 'Hashes match: successfully verified JBlake2 %s\n' "${version}"
	}

	main
	EOF
	chmod +x "${ENTRYPOINT_SCRIPT_NAME}"
}

build_docker_image()
{
	# We're using snapshot.debian.org repositories so that we can install the exact versions used to release the
	# project. See https://snapshot.debian.org for more information.
	# Though not required (since we rely on snapshot.debian.org), versions of key packages are explicitly specified
	# for documentation purposes. Note: on Debian systems, you can use 'apt list -a <package>' to list all
	# available versions of a particular package.
	docker build --quiet --tag "${BUILD_IMAGE_NAME}" . -f -<<- EOF
	FROM "${BASE_IMAGE_NAME}"
	RUN printf 'deb http://snapshot.debian.org/archive/debian/20230908T030331Z/ bullseye main' > /etc/apt/sources.list \
		&& apt update && apt install -y --no-install-recommends \
		curl \
		maven=3.6.3-5 \
		openjdk-17-jdk-headless=17.0.6+10-1~deb11u1 \
		&& rm -rf /var/lib/apt/lists/*
	COPY . /usr/src/jblake2
	WORKDIR /usr/src/jblake2
	COPY ./"${ENTRYPOINT_SCRIPT_NAME}" /
	ENTRYPOINT ["/${ENTRYPOINT_SCRIPT_NAME}"]
	EOF
}

run_build_container()
{
	docker run --rm "${BUILD_IMAGE_NAME}"
}

cleanup()
{
	local -r result=${?}
	docker image rm "${BUILD_IMAGE_NAME}" "${BASE_IMAGE_NAME}" > /dev/null 2>&1 || true
	rm "${ENTRYPOINT_SCRIPT_NAME}" > /dev/null 2>&1 || true
	exit ${result}
}

main()
{
	trap cleanup EXIT ERR
	printf 'Checking prerequisites...\t'
	check_bash_version
	check_cpu_architecture
	check_docker_version
	printf 'OK\n'
	printf 'Building docker image...\t'
	write_entrypoint
	build_docker_image
	printf 'Running build container...\n'
	run_build_container
}

main


src/org/kocakosm/jblake2/Blake2.java
/*----------------------------------------------------------------------------*
 * This file is part of JBlake2.                                              *
 * Copyright (C) Osman Koçak <kocakosm@gmail.com>                             *
 *                                                                            *
 * This program is free software: you can redistribute it and/or modify it    *
 * under the terms of the GNU Lesser General Public License as published by   *
 * the Free Software Foundation, either version 3 of the License, or (at your *
 * option) any later version.                                                 *
 * This program is distributed in the hope that it will be useful, but        *
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY *
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public     *
 * License for more details.                                                  *
 * You should have received a copy of the GNU Lesser General Public License   *
 * along with this program. If not, see <http://www.gnu.org/licenses/>.       *
 *----------------------------------------------------------------------------*/

package org.kocakosm.jblake2;

/**
 * Common interface for BLAKE2 flavors.
 *
 * @see Blake2b
 * @see Blake2s
 * @see <a href="https://blake2.net">blake2.net</a>
 * @see <a href="https://tools.ietf.org/html/rfc7693">RFC 7693</a>
 *
 * @author Osman Koçak
 */
public interface Blake2
{
	/**
	 * Returns the algorithm name.
	 *
	 * @return the algorithm name.
	 */
	String algorithm();

	/**
	 * Returns the digest's length (in bytes).
	 *
	 * @return the digest's length (in bytes).
	 */
	int length();

	/**
	 * Returns a copy of this object.
	 *
	 * @return a copy of this object.
	 */
	Blake2 copy();

	/**
	 * Resets the digest. Note that this does not erase the key.
	 *
	 * @return this object.
	 */
	Blake2 reset();

	/**
	 * Erases the key and resets the digest, thus making this instance
	 * functionally equivalent to a newly created unkeyed digest. If this
	 * instance is already unkeyed, calling this method has no other effect
	 * than resetting the digest.
	 * Please note that this method can only guarantee that the active copy
	 * of the key is erased from memory. Indeed, in practice, an object may
	 * have several copies of itself in physical RAM (garbage collection
	 * algorithms tend to regularly "move" objects in memory).
	 *
	 * @return this object.
	 */
	Blake2 burn();

	/**
	 * Updates the digest using the given byte.
	 *
	 * @param input the byte with which to update the digest.
	 *
	 * @return this object.
	 */
	Blake2 update(byte input);

	/**
	 * Updates the digest using the given array of bytes.
	 *
	 * @param input the array of bytes with which to update the digest.
	 *
	 * @return this object.
	 *
	 * @throws NullPointerException if {@code input} is {@code null}.
	 */
	Blake2 update(byte... input);

	/**
	 * Updates the digest using the specified number of bytes from the given
	 * array, starting at the specified offset.
	 *
	 * @param input the array of bytes.
	 * @param off the offset to start from in the array, inclusive.
	 * @param len the number of bytes to use, starting at {@code off}.
	 *
	 * @return this object.
	 *
	 * @throws NullPointerException if {@code input} is {@code null}.
	 * @throws IllegalArgumentException if {@code off < 0 || len < 0}.
	 * @throws IndexOutOfBoundsException if {@code off + len > input.length}.
	 */
	Blake2 update(byte[] input, int off, int len);

	/**
	 * Performs a final update on the digest using the specified array of
	 * bytes, then completes the digest computation. That is, this method
	 * first calls {@link #update(byte...)}, passing the input array to the
	 * update method, then calls {@link #digest()}. Note that the digest is
	 * reset after this call is made.
	 *
	 * @param input the array of bytes with which to update the digest
	 *	before completing its computation.
	 *
	 * @return the resulting digest.
	 *
	 * @throws NullPointerException if {@code input} is {@code null}.
	 */
	default byte[] digest(byte... input)
	{
		return digest(input, 0, input.length);
	}

	/**
	 * Performs a final update on the digest using the specified data bytes,
	 * then completes the digest computation. That is, this method first
	 * calls {@link #update(byte[], int, int)}, passing the input array to
	 * the update method, then calls {@link #digest()}. Note that the digest
	 * is reset after this call is made.
	 *
	 * @param input the array of bytes.
	 * @param off the offset to start from in the array, inclusive.
	 * @param len the number of bytes to use, starting at {@code off}.
	 *
	 * @return the resulting digest.
	 *
	 * @throws NullPointerException if {@code input} is {@code null}.
	 * @throws IllegalArgumentException if {@code off < 0 || len < 0}.
	 * @throws IndexOutOfBoundsException if {@code off + len > input.length}.
	 */
	default byte[] digest(byte[] input, int off, int len)
	{
		return update(input, off, len).digest();
	}

	/**
	 * Completes the digest computation. Note that this digest is reset
	 * after this call is made.
	 *
	 * @return the resulting digest.
	 */
	byte[] digest();
}

src/org/kocakosm/jblake2/Preconditions.java
/*----------------------------------------------------------------------------*
 * This file is part of JBlake2.                                              *
 * Copyright (C) Osman Koçak <kocakosm@gmail.com>                             *
 *                                                                            *
 * This program is free software: you can redistribute it and/or modify it    *
 * under the terms of the GNU Lesser General Public License as published by   *
 * the Free Software Foundation, either version 3 of the License, or (at your *
 * option) any later version.                                                 *
 * This program is distributed in the hope that it will be useful, but        *
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY *
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public     *
 * License for more details.                                                  *
 * You should have received a copy of the GNU Lesser General Public License   *
 * along with this program. If not, see <http://www.gnu.org/licenses/>.       *
 *----------------------------------------------------------------------------*/

package org.kocakosm.jblake2;

/**
 * Static utility methods to check whether preconditions required for the
 * execution of a method are met.
 *
 * @author Osman Koçak
 */
final class Preconditions
{
	/**
	 * Checks the truth of the given condition checking arguments validity.
	 *
	 * @param condition the boolean condition to test.
	 *
	 * @throws IllegalArgumentException if the condition is {@code false}.
	 */
	static void checkArgument(boolean condition)
	{
		if (!condition) {
			throw new IllegalArgumentException();
		}
	}

	/**
	 * Checks that the specified range is valid and within the given array's
	 * bounds.
	 *
	 * @param buf the array.
	 * @param off the offset of the range in the array, inclusive.
	 * @param len the length of the range, starting at {@code off}.
	 *
	 * @throws IllegalArgumentException if {@code off < 0 || len < 0}.
	 * @throws IndexOutOfBoundsException if {@code off + len > buf.length}.
	 */
	static void checkBounds(byte[] buf, int off, int len)
	{
		checkArgument(off >= 0 && len >= 0);
		if (off + len > buf.length) {
			throw new IndexOutOfBoundsException();
		}
	}

	/**
	 * Checks the truth of the given condition checking the state of the
	 * calling instance.
	 *
	 * @param condition the boolean condition to test.
	 *
	 * @throws IllegalStateException if the condition is {@code false}.
	 */
	static void checkState(boolean condition)
	{
		if (!condition) {
			throw new IllegalStateException();
		}
	}

	private Preconditions()
	{
		// See "Effective Java" (Item 4)
		throw new AssertionError("Not meant to be instantiated");
	}
}

src/org/kocakosm/jblake2/Blake2b.java
/*----------------------------------------------------------------------------*
 * This file is part of JBlake2.                                              *
 * Copyright (C) Osman Koçak <kocakosm@gmail.com>                             *
 *                                                                            *
 * This program is free software: you can redistribute it and/or modify it    *
 * under the terms of the GNU Lesser General Public License as published by   *
 * the Free Software Foundation, either version 3 of the License, or (at your *
 * option) any later version.                                                 *
 * This program is distributed in the hope that it will be useful, but        *
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY *
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public     *
 * License for more details.                                                  *
 * You should have received a copy of the GNU Lesser General Public License   *
 * along with this program. If not, see <http://www.gnu.org/licenses/>.       *
 *----------------------------------------------------------------------------*/

package org.kocakosm.jblake2;

import java.util.Arrays;

/**
 * The BLAKE2b digest algorithm. This BLAKE2 flavor is highly optimized for
 * 64-bit platforms and produces digests of any size between 1 and 64 bytes.
 * This class implements the BLAKE2b algorithm as specified in RFC 7693. The
 * <a href="https://blake2.net/blake2.pdf">original paper</a> defines some
 * additional variants with features such as salting, personalization and tree
 * hashing. These features are considered optional and not covered by the RFC.
 * BLAKE2b can be directly keyed, making it functionally equivalent to a Message
 * Authentication Code (it does not require a special construction like HMAC).
 * Instances of this class are not thread safe.
 *
 * @see <a href="https://blake2.net">blake2.net</a>
 * @see <a href="https://tools.ietf.org/html/rfc7693">RFC 7693</a>
 *
 * @author Osman Koçak
 */
public final class Blake2b implements Blake2
{
	private static final int BLOCK_LENGTH = 128;
	private static final long IV[] = {
		0X6A09E667F3BCC908L, 0XBB67AE8584CAA73BL, 0X3C6EF372FE94F82BL,
		0XA54FF53A5F1D36F1L, 0X510E527FADE682D1L, 0X9B05688C2B3E6C1FL,
		0X1F83D9ABFB41BD6BL, 0X5BE0CD19137E2179L
	};
	private static final int[][] SIGMA = {
		{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15},
		{14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3},
		{11, 8, 12, 0, 5, 2, 15, 13, 10, 14, 3, 6, 7, 1, 9, 4},
		{7, 9, 3, 1, 13, 12, 11, 14, 2, 6, 5, 10, 4, 0, 15, 8},
		{9, 0, 5, 7, 2, 4, 10, 15, 14, 1, 11, 12, 6, 8, 3, 13},
		{2, 12, 6, 10, 0, 11, 8, 3, 4, 13, 7, 5, 15, 14, 1, 9},
		{12, 5, 1, 15, 14, 13, 4, 10, 0, 7, 6, 3, 9, 2, 8, 11},
		{13, 11, 7, 14, 12, 1, 3, 9, 5, 0, 15, 4, 8, 6, 2, 10},
		{6, 15, 14, 9, 11, 3, 0, 8, 12, 2, 13, 7, 1, 4, 10, 5},
		{10, 2, 8, 4, 7, 6, 1, 5, 15, 11, 9, 14, 3, 12, 13, 0},
		{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15},
		{14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3}
	};

	private final int digestLength;
	private final byte[] buffer;
	private byte[] key;
	private long[] h; // internal state
	private long t0; // counter's LSB
	private long t1; // counter's MSB
	private int c; // number of bytes in the buffer

	/**
	 * Creates a new unkeyed {@code Blake2b} instance.
	 *
	 * @param digestLength the desired digest's length (in bytes).
	 *
	 * @throws IllegalArgumentException if {@code digestLength} is not in
	 *	the {@code [1, 64]} range.
	 */
	public Blake2b(int digestLength)
	{
		this(digestLength, new byte[0]);
	}

	/**
	 * Creates a new keyed {@code Blake2b} instance. If the key is empty,
	 * then the created instance is equivalent to an unkeyed digest. The
	 * given key can be safely erased from memory after this constructor has
	 * been called.
	 *
	 * @param digestLength the desired digest's length (in bytes).
	 * @param key the key to use.
	 *
	 * @throws NullPointerException if {@code key} is {@code null}.
	 * @throws IllegalArgumentException if {@code key}'s length is greater
	 *	than {@code 64} or if {@code digestLength} is not in the
	 *	{@code [1, 64]} range.
	 */
	public Blake2b(int digestLength, byte[] key)
	{
		Preconditions.checkArgument(key.length <= 64);
		Preconditions.checkArgument(digestLength >= 1 && digestLength <= 64);
		this.buffer = new byte[BLOCK_LENGTH];
		this.key = Arrays.copyOf(key, key.length);
		this.digestLength = digestLength;
		reset();
	}

	private Blake2b(Blake2b digest)
	{
		this.c = digest.c;
		this.buffer = Arrays.copyOf(digest.buffer, digest.buffer.length);
		this.key = Arrays.copyOf(digest.key, digest.key.length);
		this.digestLength = digest.digestLength;
		this.h = Arrays.copyOf(digest.h, digest.h.length);
		this.t0 = digest.t0;
		this.t1 = digest.t1;
	}

	@Override
	public String algorithm()
	{
		return "BLAKE2b";
	}

	@Override
	public int length()
	{
		return digestLength;
	}

	@Override
	public Blake2b copy()
	{
		return new Blake2b(this);
	}

	@Override
	public Blake2b reset()
	{
		t0 = 0L;
		t1 = 0L;
		h = Arrays.copyOf(IV, IV.length);
		h[0] ^= digestLength | (key.length << 8) | 0x01010000;
		if (key.length > 0) {
			System.arraycopy(key, 0, buffer, 0, key.length);
			Arrays.fill(buffer, key.length, BLOCK_LENGTH, (byte) 0);
			c = BLOCK_LENGTH;
		} else {
			c = 0;
		}
		return this;
	}

	@Override
	public Blake2b burn()
	{
		Arrays.fill(key, (byte) 0);
		Arrays.fill(buffer, (byte) 0); // buffer may contain the key...
		key = new byte[0];
		reset();
		return this;
	}

	@Override
	public Blake2b update(byte input)
	{
		if (c == BLOCK_LENGTH) {
			processBuffer(false);
		}
		buffer[c++] = input;
		return this;
	}

	@Override
	public Blake2b update(byte... input)
	{
		return update(input, 0, input.length);
	}

	@Override
	public Blake2b update(byte[] input, int off, int len)
	{
		Preconditions.checkBounds(input, off, len);
		int index = off;
		int remaining = len;
		while (remaining > 0) {
			if (c == BLOCK_LENGTH) {
				processBuffer(false);
			}
			int cpLen = Math.min(BLOCK_LENGTH - c, remaining);
			System.arraycopy(input, index, buffer, c, cpLen);
			remaining -= cpLen;
			index += cpLen;
			c += cpLen;
		}
		return this;
	}

	@Override
	public byte[] digest()
	{
		Arrays.fill(buffer, c, BLOCK_LENGTH, (byte) 0);
		processBuffer(true);
		byte[] out = new byte[digestLength];
		int i = 0;
		while (i * 8 < digestLength - 8) {
			LittleEndian.encode(h[i], out, i * 8);
			i++;
		}
		byte[] last = LittleEndian.encode(h[i]);
		System.arraycopy(last, 0, out, i * 8, digestLength - (i * 8));
		reset();
		return out;
	}

	private void processBuffer(boolean lastBlock)
	{
		t0 += c;
		if (t0 == 0L && c > 0) { // bitwise overflow
			t1++;
			Preconditions.checkState(t1 != 0L);
		}
		c = 0;
		F(buffer, lastBlock);
	}

	private void F(byte[] input, boolean lastBlock)
	{
		long[] v = new long[16];
		System.arraycopy(h, 0, v, 0, h.length);
		System.arraycopy(IV, 0, v, h.length, IV.length);
		v[12] ^= t0;
		v[13] ^= t1;
		if (lastBlock) {
			v[14] = ~v[14];
		}
		long[] m = new long[16];
		for (int j = 0; j < 16; j++) {
			m[j] = LittleEndian.decodeLong(input, j * 8);
		}
		for (int i = 0; i < 12; i++) {
			G(v, 0, 4,  8, 12, m[SIGMA[i][ 0]], m[SIGMA[i][ 1]]);
			G(v, 1, 5,  9, 13, m[SIGMA[i][ 2]], m[SIGMA[i][ 3]]);
			G(v, 2, 6, 10, 14, m[SIGMA[i][ 4]], m[SIGMA[i][ 5]]);
			G(v, 3, 7, 11, 15, m[SIGMA[i][ 6]], m[SIGMA[i][ 7]]);
			G(v, 0, 5, 10, 15, m[SIGMA[i][ 8]], m[SIGMA[i][ 9]]);
			G(v, 1, 6, 11, 12, m[SIGMA[i][10]], m[SIGMA[i][11]]);
			G(v, 2, 7,  8, 13, m[SIGMA[i][12]], m[SIGMA[i][13]]);
			G(v, 3, 4,  9, 14, m[SIGMA[i][14]], m[SIGMA[i][15]]);
		}
		for (int i = 0; i < h.length; i++) {
			h[i] ^= v[i] ^ v[i + 8];
		}
	}

	private void G(long[] v, int a, int b, int c, int d, long x, long y)
	{
		v[a] += v[b] + x;
		v[d] = Long.rotateRight(v[d] ^ v[a], 32);
		v[c] += v[d];
		v[b] = Long.rotateRight(v[b] ^ v[c], 24);
		v[a] += v[b] + y;
		v[d] = Long.rotateRight(v[d] ^ v[a], 16);
		v[c] += v[d];
		v[b] = Long.rotateRight(v[b] ^ v[c], 63);
	}
}